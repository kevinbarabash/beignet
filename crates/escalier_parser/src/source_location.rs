use std::cmp::{max, min};
use std::ops::Range;

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Position {
    pub line: usize,
    pub column: usize,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct SourceLocation {
    pub start: Position,
    pub end: Position,
}

pub type Span = Range<usize>;

pub const DUMMY_SPAN: Span = Span { start: 0, end: 0 };

pub fn merge_spans(left: &Span, right: &Span) -> Span {
    Span {
        start: min(left.start, right.start),
        end: max(left.end, right.end),
    }
}
