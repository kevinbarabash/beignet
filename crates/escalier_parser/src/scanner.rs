use crate::source_location::Position;

#[derive(Clone)]
pub struct Scanner<'a> {
    cursor: usize,
    column: usize,
    line: usize,
    input: &'a str,
    // characters: Vec<char>,
}

impl<'a> Scanner<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            cursor: 0,
            column: 1,
            line: 1,
            input,
            // characters: string.chars().collect(),
        }
    }

    /// Returns the current cursor. Useful for reporting errors.
    pub fn cursor(&self) -> usize {
        self.cursor
    }

    pub fn position(&self) -> Position {
        Position {
            line: self.line,
            column: self.column,
        }
    }

    /// Returns the next character without advancing the cursor.
    /// AKA "lookahead"
    pub fn peek(&self, lookahead: usize) -> Option<char> {
        let start = self.cursor + lookahead;
        let end = start + 1;
        self.input
            .get(start..end)
            .map(|sub_str| sub_str.chars().next().unwrap())
    }

    /// Returns true if further progress is not possible.
    pub fn is_done(&self) -> bool {
        self.cursor == self.input.len()
    }

    /// Returns the next character (if available) and advances the cursor.
    pub fn pop(&mut self) -> Option<char> {
        let start = self.cursor;
        let end = start + 1;
        match self.input.get(start..end) {
            Some(str) => {
                self.cursor += 1;
                if str == "\n" {
                    self.line += 1;
                    self.column = 1;
                } else {
                    self.column += 1;
                }

                str.get(0..1).map(|sub_str| sub_str.chars().next().unwrap())
            }
            None => None,
        }
    }

    // pub fn backup(&mut self) ->
}
