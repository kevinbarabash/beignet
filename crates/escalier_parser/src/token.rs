use crate::expr::Expr;
use crate::jsx::{JSXElement, JSXFragment};
use crate::source_location::SourceLocation;

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum TokenKind {
    Identifier(String), // [a-zA-Z_][a-zA-Z0-9_]*

    // Literals
    BoolLit(bool),
    NumLit(String),
    StrLit(String),
    StrTemplateLit {
        parts: Vec<Token>, // This should only contain StrLit tokens
        exprs: Vec<Expr>,
    },
    Null,
    Undefined,
    JSXElement(JSXElement),
    JSXFragment(JSXFragment),

    // Types
    Number,
    Boolean,
    String,
    Symbol,

    // Keywords
    Let,
    Return,
    Fn,
    Await,
    Async,
    If,
    Else,
    Mut, // denotes a binding to a mutable reference
    Var, // denotes a re-assignable binding
    Match,
    Try,
    Catch,
    Finally,
    Do,

    // Binary Operators
    Plus,
    Minus,
    Times,
    Divide,
    Modulo,
    Equals,
    NotEquals,
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
    Or,
    And,

    // Assignment Operators
    Assign,
    PlusAssign,
    MinusAssign,
    TimesAssign,
    DivideAssign,
    ModuloAssign,

    // punctuation
    Semicolon,
    Comma,
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
    Arrow,
    Underscore,
    Question,
    QuestionDot, // used for optional chaining
    Dot,
    DotDot,    // used for ranges
    DotDotDot, // used for rest/spread
    Colon,

    Eof,
}

#[derive(Debug, PartialEq, Eq, Clone)]

pub struct Token {
    pub kind: TokenKind,
    pub loc: SourceLocation,
}
