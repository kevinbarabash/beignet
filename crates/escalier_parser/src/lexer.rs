use core::panic;

use crate::scanner::Scanner;
use crate::source_location::*;
use crate::token::*;

pub struct Lexer {
    scanner: Scanner,
}

impl Lexer {
    pub fn new(string: &str) -> Self {
        Self {
            scanner: Scanner::new(string),
        }
    }

    pub fn lex(&mut self) -> Vec<Token> {
        let mut tokens = Vec::new();
        while !self.scanner.is_done() {
            let character = self.scanner.peek(0).unwrap();
            let start = self.scanner.position();
            let kind = match character {
                'a'..='z' | 'A'..='Z' | '_' => {
                    tokens.push(self.lex_ident(start));
                    continue;
                }
                '0'..='9' => {
                    tokens.push(self.lex_number(start));
                    continue;
                }
                '=' => match self.scanner.peek(1) {
                    Some('=') => {
                        self.scanner.pop();
                        TokenKind::Equals
                    }
                    _ => TokenKind::Assign,
                },
                '+' => TokenKind::Plus,
                '-' => TokenKind::Minus,
                '(' => TokenKind::LeftParen,
                ')' => TokenKind::RightParen,
                '{' => TokenKind::LeftBrace,
                '}' => TokenKind::RightBrace,
                '[' => TokenKind::LeftBracket,
                ']' => TokenKind::RightBracket,
                ',' => TokenKind::Comma,
                '.' => TokenKind::Dot,
                ';' => TokenKind::Semicolon,
                ':' => TokenKind::Colon,
                '?' => TokenKind::Question,
                '<' => match self.scanner.peek(1) {
                    Some('=') => {
                        self.scanner.pop();
                        TokenKind::LessThanOrEqual
                    }
                    _ => TokenKind::LessThan,
                },
                '>' => match self.scanner.peek(1) {
                    Some('=') => {
                        self.scanner.pop();
                        TokenKind::GreaterThanOrEqual
                    }
                    _ => TokenKind::GreaterThan,
                },
                '!' => match self.scanner.peek(1) {
                    Some('=') => {
                        self.scanner.pop();
                        TokenKind::NotEquals
                    }
                    _ => panic!("Unexpected character: '{}'", character),
                },
                _ => {
                    // TODO: Error handling
                    self.scanner.pop();
                    continue;
                }
            };
            self.scanner.pop();
            tokens.push(Token {
                kind,
                loc: SourceLocation {
                    start,
                    end: self.scanner.position(),
                },
            });
        }
        tokens
    }

    pub fn lex_ident(&mut self, start: Position) -> Token {
        let mut ident = String::new();
        while !self.scanner.is_done() {
            let character = self.scanner.peek(0).unwrap();
            match character {
                'a'..='z' | 'A'..='Z' | '_' | '0'..='9' => {
                    ident.push(*character);
                    self.scanner.pop();
                }
                _ => {
                    break;
                }
            }
        }
        Token {
            kind: TokenKind::Identifier(ident),
            loc: SourceLocation {
                start,
                end: self.scanner.position(),
            },
        }
    }

    pub fn lex_number(&mut self, start: Position) -> Token {
        let mut number = String::new();
        let mut decimal = false;
        while !self.scanner.is_done() {
            let character = self.scanner.peek(0).unwrap();
            match character {
                '0'..='9' => {
                    number.push(*character);
                    self.scanner.pop();
                }
                '.' => {
                    if decimal {
                        panic!("Unexpected character: '{}'", character);
                    }
                    number.push(*character);
                    self.scanner.pop();
                    decimal = true;
                }
                _ => {
                    break;
                }
            }
        }
        Token {
            kind: TokenKind::Number(number),
            loc: SourceLocation {
                start,
                end: self.scanner.position(),
            },
        }
    }
}
