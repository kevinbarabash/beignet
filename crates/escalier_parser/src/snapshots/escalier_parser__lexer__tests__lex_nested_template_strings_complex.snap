---
source: crates/escalier_parser/src/lexer.rs
expression: "tokens[0].kind"
---
StrTemplateLit {
    parts: [
        Token {
            kind: StrLit(
                "ids = ",
            ),
            loc: SourceLocation {
                start: Position {
                    line: 1,
                    column: 1,
                },
                end: Position {
                    line: 1,
                    column: 8,
                },
            },
        },
        Token {
            kind: StrLit(
                "ids.map(fn (id) => ",
            ),
            loc: SourceLocation {
                start: Position {
                    line: 1,
                    column: 10,
                },
                end: Position {
                    line: 1,
                    column: 30,
                },
            },
        },
    ],
    exprs: [
        Expr {
            kind: Call {
                args: [
                    Expr {
                        kind: Function {
                            params: [
                                FuncParam {
                                    pattern: Pattern {
                                        loc: SourceLocation {
                                            start: Position {
                                                line: 1,
                                                column: 22,
                                            },
                                            end: Position {
                                                line: 1,
                                                column: 24,
                                            },
                                        },
                                        kind: Ident(
                                            BindingIdent {
                                                name: "id",
                                                loc: SourceLocation {
                                                    start: Position {
                                                        line: 1,
                                                        column: 22,
                                                    },
                                                    end: Position {
                                                        line: 1,
                                                        column: 24,
                                                    },
                                                },
                                                mutable: false,
                                            },
                                        ),
                                    },
                                    type_ann: None,
                                    optional: false,
                                },
                            ],
                            body: Expr(
                                Expr {
                                    kind: TemplateLiteral {
                                        parts: [
                                            String(
                                                "x",
                                            ),
                                            String(
                                                "id}",
                                            ),
                                        ],
                                        exprs: [
                                            Expr {
                                                kind: Identifier(
                                                    "id",
                                                ),
                                                loc: SourceLocation {
                                                    start: Position {
                                                        line: 1,
                                                        column: 33,
                                                    },
                                                    end: Position {
                                                        line: 1,
                                                        column: 35,
                                                    },
                                                },
                                            },
                                        ],
                                    },
                                    loc: SourceLocation {
                                        start: Position {
                                            line: 1,
                                            column: 28,
                                        },
                                        end: Position {
                                            line: 1,
                                            column: 37,
                                        },
                                    },
                                },
                            ),
                            type_ann: None,
                        },
                        loc: SourceLocation {
                            start: Position {
                                line: 1,
                                column: 18,
                            },
                            end: Position {
                                line: 1,
                                column: 37,
                            },
                        },
                    },
                ],
                callee: Expr {
                    kind: Member {
                        object: Expr {
                            kind: Identifier(
                                "ids",
                            ),
                            loc: SourceLocation {
                                start: Position {
                                    line: 1,
                                    column: 10,
                                },
                                end: Position {
                                    line: 1,
                                    column: 13,
                                },
                            },
                        },
                        property: Expr {
                            kind: Identifier(
                                "map",
                            ),
                            loc: SourceLocation {
                                start: Position {
                                    line: 1,
                                    column: 14,
                                },
                                end: Position {
                                    line: 1,
                                    column: 17,
                                },
                            },
                        },
                    },
                    loc: SourceLocation {
                        start: Position {
                            line: 1,
                            column: 10,
                        },
                        end: Position {
                            line: 1,
                            column: 17,
                        },
                    },
                },
            },
            loc: SourceLocation {
                start: Position {
                    line: 1,
                    column: 10,
                },
                end: Position {
                    line: 1,
                    column: 38,
                },
            },
        },
    ],
}
